src/clr-angular/data/datagrid/interfaces/filter.state.interface.ts

```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+export interface FilterStateInterface {
```

Comment:
I think this needs to be specific for each filter. There are filters that span multiple properties or that do not have a string value. Each filter therefore needs to provide it's own filter interface and the `compatibleToState` method needs to accept `any`.

Reply:
Or maybe, instead of `any`, just a very basic interface with only the type which state interface of the individual interfaces then extend. We always lose the type safety anyway when serializing the stuff.

Reply:
I think every filter has a type, a property name and the value itself, so we should not remove any of these just we should go a generic type parameter, which gives the type of the value, and it should be the same type which is given in the type property in string format.
Pseudo code:
`FilterStateInterface <T>{
    type: string; //if T is Date then this value should be "Date", so we can preserve the type after serialization
    property: string;
    value: T;
}`

What's your opinion?

Reply:
I don't think that every filter has a property and the value of a filter is not usually a `T`, because `T` is the value of the row and not the column.

```
FilterStateInterface {
    readonly type: string
}
```
is the only thing I would require and then the builtin string would have it's own interface:
```
StringFilterStateInterface extends FilterStateInterface {
    property: string,
    value: string
}
```

You can then cast the state here:
```
compatibleToState(state: any): boolean {
    if nstate.type === 'BuiltinStringFilter') {
         return state.property === this.property
     }
   else {
        return false
   }
}
```
This is semantically the same as my code snippet in https://github.com/Qabel/clarity/pull/1#issuecomment-442863392 , I just unrolled the code.

Reply:
It's pushed now. I think instead of any we need FilterStateInterface. Please check the last commit.